File: polymarket_api.py
"""Polymarket API client for fetching whale transactions."""

import requests
import time
from datetime import datetime, timedelta
from typing import List, Dict, Optional
import config


class PolymarketAPI:
    """Client for interacting with Polymarket Data API."""
    
    def __init__(self):
        """Initialize API client."""
        self.base_url = config.POLYMARKET_API_BASE
        self.trades_endpoint = config.TRADES_ENDPOINT
        self.timeout = config.API_TIMEOUT
        self.max_retries = config.MAX_RETRIES
        
    def fetch_trades(
        self,
        start_time: Optional[int] = None,
        end_time: Optional[int] = None,
        limit: int = config.TRADES_LIMIT
    ) -> List[Dict]:
        """
        Fetch whale trades from Polymarket API.
        
        Args:
            start_time: Start timestamp in seconds (optional)
            end_time: End timestamp in seconds (optional)
            limit: Maximum number of trades to fetch
            
        Returns:
            List of trade dictionaries
        """
        params = {
            'filterType': config.FILTER_TYPE,
            'filterAmount': config.FILTER_AMOUNT,
            'limit': limit,
            'sortBy': 'TIMESTAMP',
            'sortDirection': 'DESC'
        }
        
        if start_time:
            params['start'] = start_time
        if end_time:
            params['end'] = end_time
            
        # Retry logic
        for attempt in range(self.max_retries):
            try:
                response = requests.get(
                    self.trades_endpoint,
                    params=params,
                    timeout=self.timeout
                )
                response.raise_for_status()
                
                data = response.json()
                
                # Parse and normalize the response
                trades = self._parse_trades(data)
                return trades
                
            except requests.exceptions.RequestException as e:
                print(f"API request failed (attempt {attempt + 1}/{self.max_retries}): {e}")
                if attempt < self.max_retries - 1:
                    time.sleep(2 ** attempt)  # Exponential backoff
                else:
                    raise
                    
        return []
        
    def _parse_trades(self, data: List[Dict]) -> List[Dict]:
        """
        Parse and normalize trade data from API response.
        
        Args:
            data: Raw API response data
            
        Returns:
            List of normalized trade dictionaries
        """
        trades = []
        
        for trade in data:
            try:
                # Extract relevant fields
                normalized_trade = {
                    'tx_hash': trade.get('transactionHash', trade.get('id', '')),
                    'amount': float(trade.get('price', 0)) * float(trade.get('size', 0)),
                    'market_name': trade.get('title', 'Unknown Market'),  # Use 'title' field directly
                    'market_id': trade.get('eventSlug', trade.get('slug', '')),  # Use eventSlug for URL (event-level, not market-level)
                    'outcome': trade.get('outcome', ''),
                    'side': trade.get('side', 'UNKNOWN'),
                    'trader_address': trade.get('proxyWallet', trade.get('takerAddress', trade.get('makerAddress', ''))),
                    'timestamp': int(trade.get('timestamp', trade.get('matchTime', 0))),
                    'details': {
                        'price': trade.get('price'),
                        'size': trade.get('size'),
                        'fee_rate': trade.get('feeRateBps'),
                        'transaction_hash': trade.get('transactionHash'),
                        'bucket_index': trade.get('bucketIndex'),
                        'match_time': trade.get('matchTime'),
                        'slug': trade.get('slug'),
                        'event_slug': trade.get('eventSlug'),
                        'raw_data': trade  # Store complete raw data for reference
                    }
                }
                
                # Only include if we have required fields
                if normalized_trade['tx_hash'] and normalized_trade['amount'] > 0:
                    trades.append(normalized_trade)
                    
            except (KeyError, ValueError, TypeError) as e:
                print(f"Error parsing trade: {e}")
                continue
                
        return trades
        
    def fetch_initial_trades(self) -> List[Dict]:
        """
        Fetch initial trades on first run.
        Tries last 24 hours first, then falls back to 7 days if no results.
        
        Returns:
            List of trade dictionaries
        """
        now = int(datetime.now().timestamp())
        
        # Try last 24 hours
        start_24h = now - (config.INITIAL_FETCH_HOURS * 3600)
        print(f"Fetching trades from last {config.INITIAL_FETCH_HOURS} hours...")
        trades = self.fetch_trades(start_time=start_24h, end_time=now)
        
        if trades:
            print(f"Found {len(trades)} whale trades in last 24 hours")
            return trades
            
        # Fallback to 7 days
        print("No trades found in last 24 hours, trying last 7 days...")
        start_7d = now - (config.FALLBACK_FETCH_DAYS * 24 * 3600)
        trades = self.fetch_trades(start_time=start_7d, end_time=now)
        
        if trades:
            print(f"Found {len(trades)} whale trades in last 7 days")
        else:
            print("No whale trades found in last 7 days")
            
        return trades
        
    def fetch_new_trades(self, last_fetch_time: int) -> List[Dict]:
        """
        Fetch trades since the last fetch time.
        
        Args:
            last_fetch_time: Unix timestamp of last fetch
            
        Returns:
            List of new trade dictionaries
        """
        now = int(datetime.now().timestamp())
        
        print(f"Fetching new trades since {datetime.fromtimestamp(last_fetch_time)}")
        trades = self.fetch_trades(start_time=last_fetch_time, end_time=now)
        
        print(f"Found {len(trades)} new whale trades")
        return trades

File: polywhale-web/lib/polymarket-api.ts
/**
 * Polymarket API client for fetching whale transactions.
 * TypeScript port of polymarket_api.py
 */

import {
  TRADES_ENDPOINT,
  FILTER_TYPE,
  DEFAULT_WHALE_THRESHOLD,
  API_TIMEOUT,
  MAX_RETRIES,
  TRADES_LIMIT,
  INITIAL_FETCH_HOURS,
  FALLBACK_FETCH_DAYS,
} from "./config";

export interface TradeDetails {
  price?: number;
  size?: number;
  fee_rate?: number;
  transaction_hash?: string;
  bucket_index?: number;
  match_time?: number;
  slug?: string;
  event_slug?: string;
  raw_data?: Record<string, unknown>;
}

export interface WhaleTrade {
  tx_hash: string;
  amount: number;
  market_name: string;
  market_id: string;
  outcome: string;
  side: string;
  trader_address: string;
  timestamp: number;
  details: TradeDetails;
}

interface RawTrade {
  transactionHash?: string;
  id?: string;
  price?: number;
  size?: number;
  title?: string;
  eventSlug?: string;
  slug?: string;
  outcome?: string;
  side?: string;
  proxyWallet?: string;
  takerAddress?: string;
  makerAddress?: string;
  timestamp?: number;
  matchTime?: number;
  feeRateBps?: number;
  bucketIndex?: number;
}

/**
 * Sleep utility for retry backoff
 */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Fetch trades from Polymarket API
 */
export async function fetchTrades(
  whaleThreshold: number = DEFAULT_WHALE_THRESHOLD,
  startTime?: number,
  endTime?: number,
  limit: number = TRADES_LIMIT
): Promise<WhaleTrade[]> {
  const params = new URLSearchParams({
    filterType: FILTER_TYPE,
    filterAmount: whaleThreshold.toString(),
    limit: limit.toString(),
    sortBy: "TIMESTAMP",
    sortDirection: "DESC",
  });

  if (startTime) {
    params.set("start", startTime.toString());
  }
  if (endTime) {
    params.set("end", endTime.toString());
  }

  const url = `${TRADES_ENDPOINT}?${params.toString()}`;

  // Retry logic with exponential backoff
  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), API_TIMEOUT);

      const response = await fetch(url, {
        signal: controller.signal,
        headers: {
          Accept: "application/json",
        },
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        throw new Error(`API request failed: ${response.status}`);
      }

      const data: RawTrade[] = await response.json();
      return parseTrades(data);
    } catch (error) {
      console.error(
        `API request failed (attempt ${attempt + 1}/${MAX_RETRIES}):`,
        error
      );
      if (attempt < MAX_RETRIES - 1) {
        await sleep(Math.pow(2, attempt) * 1000); // Exponential backoff
      } else {
        throw error;
      }
    }
  }

  return [];
}

/**
 * Parse and normalize trade data from API response
 */
function parseTrades(data: RawTrade[]): WhaleTrade[] {
  const trades: WhaleTrade[] = [];

  for (const trade of data) {
    try {
      const txHash = trade.transactionHash || trade.id || "";
      const price = Number(trade.price) || 0;
      const size = Number(trade.size) || 0;
      const amount = price * size;

      if (!txHash || amount <= 0) {
        continue;
      }

      const normalizedTrade: WhaleTrade = {
        tx_hash: txHash,
        amount: amount,
        market_name: trade.title || "Unknown Market",
        market_id: trade.eventSlug || trade.slug || "",
        outcome: trade.outcome || "",
        side: trade.side || "UNKNOWN",
        trader_address:
          trade.proxyWallet || trade.takerAddress || trade.makerAddress || "",
        timestamp: Number(trade.timestamp) || Number(trade.matchTime) || 0,
        details: {
          price: trade.price,
          size: trade.size,
          fee_rate: trade.feeRateBps,
          transaction_hash: trade.transactionHash,
          bucket_index: trade.bucketIndex,
          match_time: trade.matchTime,
          slug: trade.slug,
          event_slug: trade.eventSlug,
          raw_data: trade as unknown as Record<string, unknown>,
        },
      };

      trades.push(normalizedTrade);
    } catch (error) {
      console.error("Error parsing trade:", error);
      continue;
    }
  }

  return trades;
}

/**
 * Fetch initial trades on first run
 */
export async function fetchInitialTrades(
  whaleThreshold: number = DEFAULT_WHALE_THRESHOLD
): Promise<WhaleTrade[]> {
  const now = Math.floor(Date.now() / 1000);

  // Try last 24 hours first
  const start24h = now - INITIAL_FETCH_HOURS * 3600;
  console.log(`Fetching trades from last ${INITIAL_FETCH_HOURS} hours...`);
  let trades = await fetchTrades(whaleThreshold, start24h, now);

  if (trades.length > 0) {
    console.log(`Found ${trades.length} whale trades in last 24 hours`);
    return trades;
  }

  // Fallback to 7 days
  console.log("No trades found in last 24 hours, trying last 7 days...");
  const start7d = now - FALLBACK_FETCH_DAYS * 24 * 3600;
  trades = await fetchTrades(whaleThreshold, start7d, now);

  if (trades.length > 0) {
    console.log(`Found ${trades.length} whale trades in last 7 days`);
  } else {
    console.log("No whale trades found in last 7 days");
  }

  return trades;
}

/**
 * Fetch trades since the last fetch time
 */
export async function fetchNewTrades(
  lastFetchTime: number,
  whaleThreshold: number = DEFAULT_WHALE_THRESHOLD
): Promise<WhaleTrade[]> {
  const now = Math.floor(Date.now() / 1000);
  console.log(
    `Fetching new trades since ${new Date(lastFetchTime * 1000).toISOString()}`
  );
  const trades = await fetchTrades(whaleThreshold, lastFetchTime, now);
  console.log(`Found ${trades.length} new whale trades`);
  return trades;
}
